### PR TITLE
mds: disallow certain file operations on "." and ".."

### PR DESCRIPTION
**Pull request by vshankar**
@batrick @ukernel 

I extended this PR to fixup return values for rmdir() and rename() on "." and ".." directories. Currently we return `-ENOENT`, but that's incorrect. While fixing that it seems that `MDCache::path_traverse()` returns `-ENOENT` for path (say) `/existingdir/.` -- so the fix (for the correct return value) ends up in the failure path. Would that be the expected behaviour?

Fixes: https://tracker.ceph.com/issues/25113
Signed-off-by: Venky Shankar <vshankar@redhat.com>

### CODE REVIEW

File: src/mds/Server.cc [hunk 1]
```diff
@@ -6306,8 +6307,10 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
   // lock
   set<SimpleLock*> rdlocks, wrlocks, xlocks;
 
-  for (int i=0; i<(int)trace.size()-1; i++)
-    rdlocks.insert(&trace[i]->lock);
+  for (auto &t : trace) {
+    rdlocks.insert(&t->lock);
```
Comment (batrick) [inline src/mds/Server.cc]: The new loop also adds `trace[trace.size()-1]` to `rdlocks`. Not quite the same loop :)

**Comment by vshankar** [comment]
> I think we also should add check to Client.cc. prevent it from sending bad request

Since it will be caught be the MDS, what would be the need to have the check in the client?

File: src/mds/CDentry.h [hunk 2]
```diff
@@ -152,6 +152,10 @@ class CDentry : public MDSCacheObject, public LRUObject, public Counter<CDentry>
   CDir *get_dir() { return dir; }
   std::string_view get_name() const { return std::string_view(name); }
 
+  bool is_dot_or_dotdot() {
+    return name == "." || name == "..";
+  }
+
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/mds/CDentry.h]: For the MDS to reject the operation,`"."` or `".."` should be the final component of the path not just a path component, correct? I think that's handled here.

File: src/mds/Server.cc (hunk 1 above)
Comment (vshankar) [reply src/mds/Server.cc]: nice catch -- guess I got a bit carried away :P

File: src/mds/CDentry.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ukernel) [reply src/mds/CDentry.h]: return error if any component of the path is "." or ".."

**Comment by ukernel** [comment]
add check to _mkdir/_symlink/_mknod/_create/_rename. make sure 'name' parameter is not . or ..

File: src/mds/CDentry.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (vshankar) [reply src/mds/CDentry.h]: why? something like "mkdir -p <mount>/existingdir1/../nonexistingdir2/nonexistingdir3" is valid, correct?

Or maybe I misunderstood the point?

[the remark this answers is not in the document]
Comment (batrick) [reply src/mds/CDentry.h]: I think it's fine to keep the check in the MDS only. It is an rare exceptional case that always results in an error. No need for duplicate handling.

[the remark this answers is not in the document]
Comment (ukernel) [reply src/mds/CDentry.h]: client should resolve any "." or ".." in the path

For "mkdir -p /existingdir1/../nonexistingdir2/nonexistingdir3"

The first step is "mkdir /existingdir1/../nonexistingdir2/".   The actual request sent to mds is "mkdir <root ino>/nonexistingdir2/"

The second step is "mkdir /existingdir1/../nonexistingdir2/nonexistingdir3" . the actul request send to mds is "mkdir <ino of /nonexistingdir2>/nonexistingdir3"

[the remark this answers is not in the document]
Comment (vshankar) [reply src/mds/CDentry.h]: > client should resolve any "." or ".." in the path
> 
> For "mkdir -p /existingdir1/../nonexistingdir2/nonexistingdir3"
> 
> The first step is "mkdir /existingdir1/../nonexistingdir2/".   The actual request sent to mds is "mkdir <root ino>/nonexistingdir2/"
> 
> The second step is "mkdir /existingdir1/../nonexistingdir2/nonexistingdir3" . the actul request send to mds is "mkdir <ino of /nonexistingdir2>/nonexistingdir3"

Right. I included that in the test case in this PR. As @batrick mentioned, having checks in the mds should suffice.

[the remark this answers is not in the document]
Comment (ukernel) [reply src/mds/CDentry.h]: I still like to avoid creating . or .. dentry

**Comment by vshankar** [comment]
@ukernel @batrick fixed and updated.

File: src/mds/Server.cc [hunk 3]
```diff
@@ -6976,6 +6997,11 @@ void Server::handle_client_rename(MDRequestRef& mdr)
     respond_to_request(mdr, -EINVAL);
     return;
   }
+  if (is_operation_on_dot_or_dotdot(srcpath)) {
+    respond_to_request(mdr, -EBUSY);
+    return;
+  }
```
Comment (ukernel) [inline src/mds/Server.cc]: destpath?

File: src/mds/Server.cc [hunk 4]
```diff
@@ -1658,6 +1658,16 @@ void Server::reply_client_request(MDRequestRef& mdr, MClientReply *reply)
...
 
+bool Server::is_operation_on_dot_or_dotdot(const filepath& path) {
+  std::string dname = path.last_dentry();
+  if (dname == "." || dname == "..") {
+    dout(20) << __func__ << ": operation is on . or .." << dendl;
+    return true;
+  }
+
+  return false;
+}
```
Comment (ukernel) [inline src/mds/Server.cc]: I think we can add is_last_dot_or_dotdot() helper to src/include/filepath.h

File: src/mds/Server.cc (hunk 3 above)
Comment (vshankar) [reply src/mds/Server.cc]: for destpath, the file would be moved to its parent directory.

File: src/mds/Server.cc (hunk 4 above)
Comment (vshankar) [reply src/mds/Server.cc]: I kept it here as it was the only place that needed the checks.

File: src/mds/Server.cc (hunk 3 above)
Comment (vshankar) [reply src/mds/Server.cc]: We need to check rename to the current directory itself and return -EBUSY (`..` is valid).

Currently -ENOENT is returned.

File: src/mds/Server.cc (hunk 4 above)
Comment (ukernel) [reply src/mds/Server.cc]: need to check path.depth()  >0

File: src/mds/Server.cc (hunk 3 above)
Comment (ukernel) [reply src/mds/Server.cc]: It does not make sense. Command 'rename  a ..' works, but the actually syscall is rename("a",  "../a")

Comment (vshankar) [reply src/mds/Server.cc]: ok -- I had the impression that ceph_rename(to, "..") would send the rename request to the mds with the appropriate "parent-inode/name" combination.

So, srcpath and destpath should be caught here, correct?

Comment (vshankar) [reply src/mds/Server.cc]: yes

**Comment by batrick** [comment]
This needs rebased to handle new message handling introduced by #22555.

File: src/mds/Server.cc (hunk 3 above)
Comment (ukernel) [reply src/mds/Server.cc]: yes, checking both srcpath and destpath here should be enough

**Comment by batrick** [comment]
retest this please

**Comment by ukernel** [comment]
LGTM

**Pull request merged**